File: src/integrations/poe.py
"""Web search integration using DuckDuckGo (free, no API key required)"""

from datetime import datetime

try:
    from ddgs import DDGS
    DDGS_AVAILABLE = True
except ImportError:
    DDGS_AVAILABLE = False
    print("⚠️  ddgs not installed. Install with: pip install ddgs")


class PoeIntegration:
    """Web search integration using DuckDuckGo (free, no API key needed)"""

    def __init__(self):
        self.search_enabled = DDGS_AVAILABLE

    async def search(self, query: str, keywords: list = None) -> dict:
        """
        Perform a web search using DuckDuckGo

        Args:
            query: Search query
            keywords: Additional keywords to enhance search

        Returns:
            Dictionary with search results
        """
        if not self.search_enabled:
            print("⚠️  DuckDuckGo search not available. Using mock search.")
            return self._mock_search(query)

        try:
            # Enhance query with keywords
            enhanced_query = query
            if keywords:
                enhanced_query = f"{query} {' '.join(keywords[:3])}"  # Limit to 3 keywords

            # Use DuckDuckGo search
            with DDGS() as ddgs:
                results = list(ddgs.text(enhanced_query, max_results=5))

            if not results:
                return self._mock_search(query)

            # Format results
            formatted_results = []
            for result in results:
                formatted_results.append({
                    'title': result.get('title', ''),
                    'url': result.get('href', ''),
                    'snippet': result.get('body', '')
                })

            # Create summary from top results
            summary = f"Found {len(formatted_results)} relevant sources about {query}. "
            if formatted_results:
                summary += f"Top result: {formatted_results[0]['snippet'][:150]}..."

            return {
                'success': True,
                'query': enhanced_query,
                'results': {
                    'summary': summary,
                    'sources': formatted_results
                },
                'timestamp': datetime.now().isoformat()
            }

        except Exception as e:
            print(f"Web search error: {e}")
            return self._mock_search(query)

    def _mock_search(self, query: str) -> dict:
        """
        Mock search results for development/testing

        Args:
            query: Search query

        Returns:
            Dictionary with mock results
        """
        return {
            'success': True,
            'query': query,
            'results': {
                'summary': f'Mock search results for: "{query}". Configure Poe API key for real search results.',
                'sources': [
                    {
                        'title': 'Example Source 1',
                        'url': 'https://example.com/1',
                        'snippet': 'This is a mock search result. Real results will appear when Poe API is configured.'
                    },
                    {
                        'title': 'Example Source 2',
                        'url': 'https://example.com/2',
                        'snippet': 'Mock data for testing purposes.'
                    }
                ]
            },
            'mock': True,
            'timestamp': datetime.now().isoformat()
        }

    async def get_news(self, topic: str) -> dict:
        """
        Get latest news or articles related to topic

        Args:
            topic: Topic to search for

        Returns:
            Dictionary with news results
        """
        query = f"latest news about {topic}"
        return await self.search(query, ['news', 'recent', 'update'])

File: src/integrations/poe.js
import axios from 'axios';
import { config } from '../../config/config.js';

class PoeIntegration {
  constructor() {
    this.apiKey = config.poe.apiKey;
    this.apiUrl = config.poe.apiUrl;
    this.botName = config.poe.botName;
  }

  /**
   * Perform a web search using Poe API
   * @param {string} query - Search query
   * @param {Array<string>} keywords - Additional keywords to enhance search
   * @returns {Promise<Object>} - Search results
   */
  async search(query, keywords = []) {
    if (!this.apiKey) {
      console.warn('Poe API key not configured. Skipping web search.');
      return {
        success: false,
        message: 'Poe API not configured',
        results: [],
      };
    }

    try {
      // Enhance query with keywords
      const enhancedQuery = keywords.length > 0
        ? `${query} ${keywords.join(' ')}`
        : query;

      // Note: Poe API structure may vary based on your access level
      // This is a generic implementation - adjust based on actual Poe API docs
      const response = await axios.post(
        `${this.apiUrl}query`,
        {
          query: enhancedQuery,
          bot: this.botName,
          api_key: this.apiKey,
        },
        {
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.apiKey}`,
          },
          timeout: 30000, // 30 second timeout
        }
      );

      return {
        success: true,
        query: enhancedQuery,
        results: response.data,
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      console.error('Poe API error:', error.message);

      // Fallback: return mock search results for development
      return this.mockSearch(query);
    }
  }

  /**
   * Mock search results for development/testing
   * @param {string} query - Search query
   * @returns {Object} - Mock results
   */
  mockSearch(query) {
    return {
      success: true,
      query,
      results: {
        summary: `Mock search results for: "${query}". Configure Poe API key for real search results.`,
        sources: [
          {
            title: 'Example Source 1',
            url: 'https://example.com/1',
            snippet: 'This is a mock search result. Real results will appear when Poe API is configured.',
          },
          {
            title: 'Example Source 2',
            url: 'https://example.com/2',
            snippet: 'Mock data for testing purposes.',
          },
        ],
      },
      mock: true,
      timestamp: new Date().toISOString(),
    };
  }

  /**
   * Get latest news or articles related to query
   * @param {string} topic - Topic to search for
   * @returns {Promise<Object>} - News results
   */
  async getNews(topic) {
    const query = `latest news about ${topic}`;
    return this.search(query, ['news', 'recent', 'update']);
  }
}

export default PoeIntegration;
